torch/_inductor/codecache.py

```diff
@@ -2096,19 +2107,18 @@ def _to_bytes(t: torch.Tensor) -> bytes:
                     return b""
 
                 if t.is_mkldnn:
-                    raw_array = ctypes.cast(
-                        torch.ops.mkldnn.data_ptr(t),
-                        ctypes.POINTER(ctypes.c_ubyte * torch.ops.mkldnn._nbytes(t)),
-                    )
-                    return bytes(raw_array.contents)
+                    data_ptr = torch.ops.mkldnn.data_ptr(t)
+                    nbytes = torch.ops.mkldnn._nbytes(t)
+                else:
+                    t_cpu = t.untyped_storage().cpu()
+                    data_ptr = t_cpu.data_ptr()
+                    nbytes = t_cpu.nbytes()
 
-                t_cpu = t.untyped_storage().cpu()
                 raw_array = ctypes.cast(
-                    t_cpu.data_ptr(),
-                    ctypes.POINTER(ctypes.c_ubyte * t_cpu.nbytes()),
+                    data_ptr,
+                    ctypes.POINTER(ctypes.c_ubyte * nbytes),
                 )
-
-                return bytes(raw_array.contents)
+                return _pad_to_alignment(bytes(raw_array.contents))
```

Comment:
Does this mean we will align to 64 bytes for CUDA as well? I don't think that's necessary?

Reply:
Since all the constants are saved into one serialized array, supposing the constants contain three tensors with the below order:
`tensor_cpu_start`, ..., `tensor_cpu_start + tensor_cpu_data_size -1`, `tensor_cuda_start`, ..., `tensor_cuda_start + tensor_cuda_data_size -1`, `tensor_mkldnn_start`, ..., `tensor_mkldnn_start + tensor_mkldnn_data_size -1`

We've set the alignment [here](https://github.com/pytorch/pytorch/blob/b054470db22a6c8ecba31c44ce54b9ca48159cdd/torch/_inductor/codecache.py#L1990) so that `tensor_cpu_start` is 64 bytes aligned.
If `tensor_cuda_data_size` is not 64 bytes, the `tensor_mkldnn_start` will start at this position: `tensor_cpu_start` + `tensor_cpu_data_size ` + `tensor_cuda_data_size ` and it won't be 64 bytes aligned.

Option B is to re-create a new buffer (which is 64 bytes aligned) for mkldnn tensors when loading the model.so instead of directly using the [buffer here](https://github.com/pytorch/pytorch/blob/b054470db22a6c8ecba31c44ce54b9ca48159cdd/torch/csrc/inductor/aoti_runtime/model.h#L210) which might not be aligned due to the above issue. This solution won't need to pad CUDA tensor during the serialization but it will introduce an extra memory copy when loading the model.

May I know if you prefer option B or any other suggestions?

Reply:
I think for majority of CUDA models, there won't be a mixed of CUDA tensor and MKLDNN tensors. You can check if all those const tensors are all CUDA tensors, and if yes, you can skip doing alignment.

Reply:
Thanks for the suggestion. Fixed as suggested.